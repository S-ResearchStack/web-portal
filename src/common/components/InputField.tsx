import React, { FC, ForwardedRef, forwardRef, ReactElement } from 'react';

import styled, { css } from 'styled-components';

import browser from 'src/common/utils/browser';
import { animation, colors, px, typography } from 'src/styles';

export const RIGHT_PADDING = 8;

interface InputFieldBaseProps extends React.InputHTMLAttributes<HTMLInputElement> {
  label?: string | ReactElement;
  helperText?: string;
  error?: React.ReactNode;
  withoutErrorText?: boolean;
  disabled?: boolean;
}
type InputFieldShellProps = {
  fixedHeight?: boolean;
  caption?: React.ReactNode;
} & React.PropsWithChildren<InputFieldBaseProps>;

type InputType = 'email' | 'password' | 'text' | 'date' | 'number';

type EndExtraProps = { component: ReactElement; extraWidth: number };

export interface InputFieldProps
  extends InputFieldBaseProps,
    React.InputHTMLAttributes<HTMLInputElement> {
  type?: InputType;
  endExtra?: EndExtraProps;
  lighten?: boolean;
  caption?: string;
}

const InputContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: ${px(8)};
  margin-top: 5px;
`;

const ExtraWrapper = styled.div<Pick<InputFieldProps, 'endExtra' | 'disabled'>>`
  > svg {
    position: relative;
    bottom: ${browser.isSafari ? px(43) : px(41)};
    left: ${({ endExtra }) => `calc(100% - ${px((endExtra?.extraWidth || 0) + RIGHT_PADDING)})`};
    path {
      fill: ${({ disabled, theme }) => disabled && theme.colors.onDisabled};
    }
  }
`;

const getDefaultBackgroundColor = ({ error, lighten }: InputFieldProps) => {
  if (error) {
    return colors.statusError10;
  }

  return colors.surface;
};

export const StyledTextField = styled.input<InputFieldProps>`
  ${typography.bodyMediumRegular};
  color: ${({ error }) => (error ? colors.statusErrorText : colors.textPrimary)};
  box-sizing: border-box;
  height: ${px(56)};
  width: 100%;
  margin: 0;
  padding: ${px(16)};
  background-color: ${getDefaultBackgroundColor};
  border: ${px(1)} solid ${colors.black08};
  border-radius: ${px(4)};
  transition: border 300ms ${animation.defaultTiming};
  caret-color: ${({ error }) => (error ? colors.statusErrorText : colors.textPrimaryBlue)};

  &:hover:enabled {
    border-color: ${({ error }) => (error ? 'transparent' : colors.primaryHovered)};
  }

  &:active:enabled,
  &:focus-visible {
    outline: none;
    border-color: ${({ error }) => (error ? 'transparent' : colors.primary)};
  }

  &:disabled {
    color: rgba(0, 0, 0, 0.38); // TODO unknown color
    border-color: transparent;
    background-color: #f8f8f8; // TODO unknown color
  }

  &::placeholder {
    color: ${colors.textSecondaryGray};
  }

  &:disabled::placeholder {
    color: ${colors.disabled};
  }

  &:-webkit-autofill,
  &:-webkit-autofill:hover,
  &:-webkit-autofill:focus {
    ${({ error, disabled, theme }) => css`
      border: ${px(1)} solid
        ${(disabled && theme.colors.disabled20) ||
        (error ? theme.colors.statusError10 : theme.colors.primary)};
      transition: background-color 5000s ease-in-out 0s;
      background-color: ${getDefaultBackgroundColor} !important;
      background-image: unset !important;
      -webkit-text-fill-color: ${colors.textDisabled};
    `};
  }

  &:-webkit-autofill:focus {
    border-color: ${({ error }) => (error ? 'transparent' : colors.primary)};
  }
`;

const InputWrapper = styled.div<
  Pick<InputFieldShellProps, 'fixedHeight'> & Pick<InputFieldProps, 'endExtra' | 'error'>
>`
  max-height: ${(p) => (p.fixedHeight ? px(56) : 'auto')};
  &:hover {
    ${StyledTextField} {
      :enabled {
        border-color: ${({ error }) => (error ? 'transparent' : colors.primaryHovered)};
      }
    }
  }
`;

interface BlockStatus {
  $disabled?: boolean;
  error?: boolean;
  required?: boolean;
}

const Label = styled.div<BlockStatus>`
  ${typography.bodyMediumSemibold};
  color: ${({ error, $disabled, theme }) =>
    (error && theme.colors.statusErrorText) ||
    ($disabled ? 'rgba(0, 0, 0, 0.38)' : theme.colors.textPrimary)}; // TODO unknown color
  height: ${px(18)};
`;

const InputDescription = styled.span<BlockStatus>`
  ${typography.bodySmallSemibold};
  color: ${({ error, $disabled, theme }) =>
    (error && theme.colors.statusErrorText) ||
    ($disabled ? 'rgba(0, 0, 0, 0.38)' : theme.colors.textPrimary)}; // TODO unknown color
  gap: ${px(8)};
  height: ${px(18)};
  width: fit-content;
  position: ${({ required }) =>
  (required ? "relative" : "static") };
  &::before {
    content: ${({ required }) => (required ? "'*'" : "''") };
    position: absolute;
    color: red;
    top: -2px;
    right: -10px;
  }
`;

const InputErrorText = styled.div<{ withOffset?: boolean }>`
  ${typography.bodySmallRegular};
  color: ${colors.statusErrorText};
  padding-left: ${({ withOffset }) => withOffset && px(16)};
  height: ${px(17)};
  margin-bottom: 10px;
`;

const InputCaption = styled.div<{ withOffset?: boolean }>`
  ${typography.labelRegular};
  color: ${colors.textSecondaryGray};
  height: ${px(17)};
  margin-top: -5px;
`;

export const InputFieldShell: FC<InputFieldShellProps> = ({
  label,
  helperText,
  error,
  className,
  disabled,
  children,
  withoutErrorText,
  fixedHeight = true,
  caption,
  required,
}) => (
  <InputContainer className={className}>
    {helperText && (
      <Label data-testid="input-label" error={!!error} $disabled={disabled}>
        {label}
      </Label>
    )}
    <InputDescription data-testid="input-description" error={!!error} $disabled={disabled} required={required}>
      {helperText || label || <>&nbsp;</>}
    </InputDescription>
    <InputWrapper error={error} fixedHeight={fixedHeight}>
      {children}
    </InputWrapper>
    {!withoutErrorText &&
      (error ? (
        <InputErrorText data-testid="input-error" withOffset={!helperText}>
          {error}
        </InputErrorText>
      ) : (
        <InputCaption withOffset={!helperText}>{caption || <>&nbsp;</>}</InputCaption>
      ))}
  </InputContainer>
);

const InputField = forwardRef(
  (
    {
      type,
      helperText,
      error,
      label,
      endExtra,
      disabled,
      className,
      withoutErrorText,
      required,
      caption,
      ...restProps
    }: InputFieldProps,
    ref: ForwardedRef<HTMLInputElement>
  ): ReactElement => (
    <InputFieldShell
      label={label}
      helperText={helperText}
      error={error}
      className={className}
      disabled={disabled}
      withoutErrorText={withoutErrorText}
      required={required}
      caption={caption}
    >
      <StyledTextField
        data-testid="input"
        ref={ref}
        error={error}
        type={type}
        disabled={disabled}
        endExtra={endExtra}
        {...restProps}
      />
      <ExtraWrapper disabled={disabled} endExtra={endExtra}>
        {endExtra?.component}
      </ExtraWrapper>
    </InputFieldShell>
  )
);

export default InputField;
