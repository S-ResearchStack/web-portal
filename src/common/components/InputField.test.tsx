import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import InputField from 'src/common/components/InputField';
import theme from 'src/styles/theme';
import { ThemeProvider } from 'styled-components/';

describe('InputField', () => {
  it('type and clear InputField test', async () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <InputField />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const input = screen.getByTestId('input');

    await userEvent.type(input, 'Samsung');
    expect(input).toHaveValue('Samsung');

    await userEvent.clear(input);
    expect(input).toHaveValue('');
  });

  it('[NEGATIVE] test InputField error state', () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <InputField error />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const input = screen.getByTestId('input');

    expect(input).toHaveStyle(`color: ${theme.colors.statusErrorText}`);
    expect(input).toHaveStyle(`caret-color: ${theme.colors.statusErrorText}`);
  });

  it('[NEGATIVE] test InputField disabled state', async () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <InputField disabled value="disabled" />
      </ThemeProvider>
    );

    const input = screen.getByTestId('input');

    expect(baseElement).toMatchSnapshot();
    expect(input).toHaveValue('disabled');
    expect(input).toHaveStyle('color: rgba(0, 0, 0, 0.38)');
    expect(input).toHaveStyle('background-color: #f8f8f8');
  });

  it('[NEGATIVE] should prevent user input while field is disabled', async () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <InputField disabled value="disabled" />
      </ThemeProvider>
    );

    const input = screen.getByTestId('input');

    await userEvent.type(input, 'Samsung');

    expect(baseElement).toMatchSnapshot();
    expect(input).not.toHaveValue('Samsung');
    expect(input).toHaveValue('disabled');
  });

  it('[NEGATIVE] should render with wrong props', async () => {
    const { baseElement, queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <InputField
          readOnly={1 as unknown as boolean}
          value={[] as unknown as string}
          label={null as unknown as string}
          error={null as unknown as string}
        />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const input = screen.getByTestId('input');
    const label = queryByTestId('input-label');
    const error = await queryByTestId('input-error');
    const description = screen.getByTestId('input-description');

    expect(input).toBeInTheDocument();
    expect(label).toBeNull();
    expect(description).toBeInTheDocument();
    expect(description).toHaveTextContent('');
    expect(error).toBeNil();
  });

  it('[NEGATIVE] should render without props', () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <InputField />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();
    expect(screen.getByTestId('input')).toBeInTheDocument();
  });
});
