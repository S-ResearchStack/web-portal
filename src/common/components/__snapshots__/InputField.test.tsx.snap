// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField [NEGATIVE] should prevent user input while field is disabled 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.5rem;
  margin-top: 5px;
}

.c5 > svg {
  position: relative;
  bottom: 2.6875rem;
  left: calc(100% - 0.5rem);
}

.c5 > svg path {
  fill: #A1A1A1;
}

.c4 {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  box-sizing: border-box;
  height: 3.5rem;
  width: 100%;
  margin: 0;
  padding: 1rem;
  background-color: #FFFFFF;
  border: 0.0625rem solid rgba(0,0,0,0.08);
  border-radius: 0.25rem;
  -webkit-transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  caret-color: #4475E3;
}

.c4:hover:enabled {
  border-color: #2F65E0;
}

.c4:active:enabled,
.c4:focus-visible {
  outline: none;
  border-color: #4475E3;
}

.c4:disabled {
  color: rgba(0,0,0,0.38);
  border-color: transparent;
  background-color: #f8f8f8;
}

.c4::-webkit-input-placeholder {
  color: #808080;
}

.c4::-moz-placeholder {
  color: #808080;
}

.c4:-ms-input-placeholder {
  color: #808080;
}

.c4::placeholder {
  color: #808080;
}

.c4:disabled::-webkit-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::-moz-placeholder {
  color: #DBDBDB;
}

.c4:disabled:-ms-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::placeholder {
  color: #DBDBDB;
}

.c4:-webkit-autofill,
.c4:-webkit-autofill:hover,
.c4:-webkit-autofill:focus {
  border: 0.0625rem solid rgba(219,219,219,0.2);
  -webkit-transition: background-color 5000s ease-in-out 0s;
  transition: background-color 5000s ease-in-out 0s;
  background-color: #FFFFFF !important;
  background-image: unset !important;
  -webkit-text-fill-color: #838383;
}

.c4:-webkit-autofill:focus {
  border-color: #4475E3;
}

.c2 {
  max-height: 3.5rem;
}

.c2:hover .c3:enabled {
  border-color: #2F65E0;
}

.c1 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 600;
  font-style: normal;
  color: rgba(0,0,0,0.38);
  gap: 0.5rem;
  height: 1.125rem;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  position: static;
}

.c1::before {
  content: '';
  position: absolute;
  color: red;
  top: -2px;
  right: -10px;
}

.c6 {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
  color: #808080;
  height: 1.0625rem;
  margin-top: -5px;
}

<body>
  <div>
    <div
      class="c0"
    >
      <span
        class="c1"
        data-testid="input-description"
      >
         
      </span>
      <div
        class="c2"
      >
        <input
          class="c3 c4"
          data-testid="input"
          disabled=""
          value="disabled"
        />
        <div
          class="c5"
          disabled=""
        />
      </div>
      <div
        class="c6"
      >
         
      </div>
    </div>
  </div>
</body>
`;

exports[`InputField [NEGATIVE] should render with wrong props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.5rem;
  margin-top: 5px;
}

.c5 > svg {
  position: relative;
  bottom: 2.6875rem;
  left: calc(100% - 0.5rem);
}

.c4 {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  box-sizing: border-box;
  height: 3.5rem;
  width: 100%;
  margin: 0;
  padding: 1rem;
  background-color: #FFFFFF;
  border: 0.0625rem solid rgba(0,0,0,0.08);
  border-radius: 0.25rem;
  -webkit-transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  caret-color: #4475E3;
}

.c4:hover:enabled {
  border-color: #2F65E0;
}

.c4:active:enabled,
.c4:focus-visible {
  outline: none;
  border-color: #4475E3;
}

.c4:disabled {
  color: rgba(0,0,0,0.38);
  border-color: transparent;
  background-color: #f8f8f8;
}

.c4::-webkit-input-placeholder {
  color: #808080;
}

.c4::-moz-placeholder {
  color: #808080;
}

.c4:-ms-input-placeholder {
  color: #808080;
}

.c4::placeholder {
  color: #808080;
}

.c4:disabled::-webkit-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::-moz-placeholder {
  color: #DBDBDB;
}

.c4:disabled:-ms-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::placeholder {
  color: #DBDBDB;
}

.c4:-webkit-autofill,
.c4:-webkit-autofill:hover,
.c4:-webkit-autofill:focus {
  border: 0.0625rem solid #4475E3;
  -webkit-transition: background-color 5000s ease-in-out 0s;
  transition: background-color 5000s ease-in-out 0s;
  background-color: #FFFFFF !important;
  background-image: unset !important;
  -webkit-text-fill-color: #838383;
}

.c4:-webkit-autofill:focus {
  border-color: #4475E3;
}

.c2 {
  max-height: 3.5rem;
}

.c2:hover .c3:enabled {
  border-color: #2F65E0;
}

.c1 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 600;
  font-style: normal;
  color: #474747;
  gap: 0.5rem;
  height: 1.125rem;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  position: static;
}

.c1::before {
  content: '';
  position: absolute;
  color: red;
  top: -2px;
  right: -10px;
}

.c6 {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
  color: #808080;
  height: 1.0625rem;
  margin-top: -5px;
}

<body>
  <div>
    <div
      class="c0"
    >
      <span
        class="c1"
        data-testid="input-description"
      >
         
      </span>
      <div
        class="c2"
      >
        <input
          class="c3 c4"
          data-testid="input"
          readonly=""
          value=""
        />
        <div
          class="c5"
        />
      </div>
      <div
        class="c6"
      >
         
      </div>
    </div>
  </div>
</body>
`;

exports[`InputField [NEGATIVE] should render without props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.5rem;
  margin-top: 5px;
}

.c5 > svg {
  position: relative;
  bottom: 2.6875rem;
  left: calc(100% - 0.5rem);
}

.c4 {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  box-sizing: border-box;
  height: 3.5rem;
  width: 100%;
  margin: 0;
  padding: 1rem;
  background-color: #FFFFFF;
  border: 0.0625rem solid rgba(0,0,0,0.08);
  border-radius: 0.25rem;
  -webkit-transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  caret-color: #4475E3;
}

.c4:hover:enabled {
  border-color: #2F65E0;
}

.c4:active:enabled,
.c4:focus-visible {
  outline: none;
  border-color: #4475E3;
}

.c4:disabled {
  color: rgba(0,0,0,0.38);
  border-color: transparent;
  background-color: #f8f8f8;
}

.c4::-webkit-input-placeholder {
  color: #808080;
}

.c4::-moz-placeholder {
  color: #808080;
}

.c4:-ms-input-placeholder {
  color: #808080;
}

.c4::placeholder {
  color: #808080;
}

.c4:disabled::-webkit-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::-moz-placeholder {
  color: #DBDBDB;
}

.c4:disabled:-ms-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::placeholder {
  color: #DBDBDB;
}

.c4:-webkit-autofill,
.c4:-webkit-autofill:hover,
.c4:-webkit-autofill:focus {
  border: 0.0625rem solid #4475E3;
  -webkit-transition: background-color 5000s ease-in-out 0s;
  transition: background-color 5000s ease-in-out 0s;
  background-color: #FFFFFF !important;
  background-image: unset !important;
  -webkit-text-fill-color: #838383;
}

.c4:-webkit-autofill:focus {
  border-color: #4475E3;
}

.c2 {
  max-height: 3.5rem;
}

.c2:hover .c3:enabled {
  border-color: #2F65E0;
}

.c1 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 600;
  font-style: normal;
  color: #474747;
  gap: 0.5rem;
  height: 1.125rem;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  position: static;
}

.c1::before {
  content: '';
  position: absolute;
  color: red;
  top: -2px;
  right: -10px;
}

.c6 {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
  color: #808080;
  height: 1.0625rem;
  margin-top: -5px;
}

<body>
  <div>
    <div
      class="c0"
    >
      <span
        class="c1"
        data-testid="input-description"
      >
         
      </span>
      <div
        class="c2"
      >
        <input
          class="c3 c4"
          data-testid="input"
        />
        <div
          class="c5"
        />
      </div>
      <div
        class="c6"
      >
         
      </div>
    </div>
  </div>
</body>
`;

exports[`InputField [NEGATIVE] test InputField disabled state 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.5rem;
  margin-top: 5px;
}

.c5 > svg {
  position: relative;
  bottom: 2.6875rem;
  left: calc(100% - 0.5rem);
}

.c5 > svg path {
  fill: #A1A1A1;
}

.c4 {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  box-sizing: border-box;
  height: 3.5rem;
  width: 100%;
  margin: 0;
  padding: 1rem;
  background-color: #FFFFFF;
  border: 0.0625rem solid rgba(0,0,0,0.08);
  border-radius: 0.25rem;
  -webkit-transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  caret-color: #4475E3;
}

.c4:hover:enabled {
  border-color: #2F65E0;
}

.c4:active:enabled,
.c4:focus-visible {
  outline: none;
  border-color: #4475E3;
}

.c4:disabled {
  color: rgba(0,0,0,0.38);
  border-color: transparent;
  background-color: #f8f8f8;
}

.c4::-webkit-input-placeholder {
  color: #808080;
}

.c4::-moz-placeholder {
  color: #808080;
}

.c4:-ms-input-placeholder {
  color: #808080;
}

.c4::placeholder {
  color: #808080;
}

.c4:disabled::-webkit-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::-moz-placeholder {
  color: #DBDBDB;
}

.c4:disabled:-ms-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::placeholder {
  color: #DBDBDB;
}

.c4:-webkit-autofill,
.c4:-webkit-autofill:hover,
.c4:-webkit-autofill:focus {
  border: 0.0625rem solid rgba(219,219,219,0.2);
  -webkit-transition: background-color 5000s ease-in-out 0s;
  transition: background-color 5000s ease-in-out 0s;
  background-color: #FFFFFF !important;
  background-image: unset !important;
  -webkit-text-fill-color: #838383;
}

.c4:-webkit-autofill:focus {
  border-color: #4475E3;
}

.c2 {
  max-height: 3.5rem;
}

.c2:hover .c3:enabled {
  border-color: #2F65E0;
}

.c1 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 600;
  font-style: normal;
  color: rgba(0,0,0,0.38);
  gap: 0.5rem;
  height: 1.125rem;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  position: static;
}

.c1::before {
  content: '';
  position: absolute;
  color: red;
  top: -2px;
  right: -10px;
}

.c6 {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
  color: #808080;
  height: 1.0625rem;
  margin-top: -5px;
}

<body>
  <div>
    <div
      class="c0"
    >
      <span
        class="c1"
        data-testid="input-description"
      >
         
      </span>
      <div
        class="c2"
      >
        <input
          class="c3 c4"
          data-testid="input"
          disabled=""
          value="disabled"
        />
        <div
          class="c5"
          disabled=""
        />
      </div>
      <div
        class="c6"
      >
         
      </div>
    </div>
  </div>
</body>
`;

exports[`InputField [NEGATIVE] test InputField error state 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.5rem;
  margin-top: 5px;
}

.c5 > svg {
  position: relative;
  bottom: 2.6875rem;
  left: calc(100% - 0.5rem);
}

.c4 {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 400;
  font-style: normal;
  color: #D14343;
  box-sizing: border-box;
  height: 3.5rem;
  width: 100%;
  margin: 0;
  padding: 1rem;
  background-color: #FFEBEB;
  border: 0.0625rem solid rgba(0,0,0,0.08);
  border-radius: 0.25rem;
  -webkit-transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  caret-color: #D14343;
}

.c4:hover:enabled {
  border-color: transparent;
}

.c4:active:enabled,
.c4:focus-visible {
  outline: none;
  border-color: transparent;
}

.c4:disabled {
  color: rgba(0,0,0,0.38);
  border-color: transparent;
  background-color: #f8f8f8;
}

.c4::-webkit-input-placeholder {
  color: #808080;
}

.c4::-moz-placeholder {
  color: #808080;
}

.c4:-ms-input-placeholder {
  color: #808080;
}

.c4::placeholder {
  color: #808080;
}

.c4:disabled::-webkit-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::-moz-placeholder {
  color: #DBDBDB;
}

.c4:disabled:-ms-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::placeholder {
  color: #DBDBDB;
}

.c4:-webkit-autofill,
.c4:-webkit-autofill:hover,
.c4:-webkit-autofill:focus {
  border: 0.0625rem solid #FFEBEB;
  -webkit-transition: background-color 5000s ease-in-out 0s;
  transition: background-color 5000s ease-in-out 0s;
  background-color: #FFEBEB !important;
  background-image: unset !important;
  -webkit-text-fill-color: #838383;
}

.c4:-webkit-autofill:focus {
  border-color: transparent;
}

.c2 {
  max-height: 3.5rem;
}

.c2:hover .c3:enabled {
  border-color: transparent;
}

.c1 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 600;
  font-style: normal;
  color: #D14343;
  gap: 0.5rem;
  height: 1.125rem;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  position: static;
}

.c1::before {
  content: '';
  position: absolute;
  color: red;
  top: -2px;
  right: -10px;
}

.c6 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 400;
  font-style: normal;
  color: #D14343;
  padding-left: 1rem;
  height: 1.0625rem;
  margin-bottom: 10px;
}

<body>
  <div>
    <div
      class="c0"
    >
      <span
        class="c1"
        data-testid="input-description"
      >
         
      </span>
      <div
        class="c2"
      >
        <input
          class="c3 c4"
          data-testid="input"
        />
        <div
          class="c5"
        />
      </div>
      <div
        class="c6"
        data-testid="input-error"
      />
    </div>
  </div>
</body>
`;

exports[`InputField type and clear InputField test 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.5rem;
  margin-top: 5px;
}

.c5 > svg {
  position: relative;
  bottom: 2.6875rem;
  left: calc(100% - 0.5rem);
}

.c4 {
  font-family: Inter,sans-serif;
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 400;
  font-style: normal;
  color: #474747;
  box-sizing: border-box;
  height: 3.5rem;
  width: 100%;
  margin: 0;
  padding: 1rem;
  background-color: #FFFFFF;
  border: 0.0625rem solid rgba(0,0,0,0.08);
  border-radius: 0.25rem;
  -webkit-transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  transition: border 300ms cubic-bezier(0.45,0.05,0,1);
  caret-color: #4475E3;
}

.c4:hover:enabled {
  border-color: #2F65E0;
}

.c4:active:enabled,
.c4:focus-visible {
  outline: none;
  border-color: #4475E3;
}

.c4:disabled {
  color: rgba(0,0,0,0.38);
  border-color: transparent;
  background-color: #f8f8f8;
}

.c4::-webkit-input-placeholder {
  color: #808080;
}

.c4::-moz-placeholder {
  color: #808080;
}

.c4:-ms-input-placeholder {
  color: #808080;
}

.c4::placeholder {
  color: #808080;
}

.c4:disabled::-webkit-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::-moz-placeholder {
  color: #DBDBDB;
}

.c4:disabled:-ms-input-placeholder {
  color: #DBDBDB;
}

.c4:disabled::placeholder {
  color: #DBDBDB;
}

.c4:-webkit-autofill,
.c4:-webkit-autofill:hover,
.c4:-webkit-autofill:focus {
  border: 0.0625rem solid #4475E3;
  -webkit-transition: background-color 5000s ease-in-out 0s;
  transition: background-color 5000s ease-in-out 0s;
  background-color: #FFFFFF !important;
  background-image: unset !important;
  -webkit-text-fill-color: #838383;
}

.c4:-webkit-autofill:focus {
  border-color: #4475E3;
}

.c2 {
  max-height: 3.5rem;
}

.c2:hover .c3:enabled {
  border-color: #2F65E0;
}

.c1 {
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.3125rem;
  font-weight: 600;
  font-style: normal;
  color: #474747;
  gap: 0.5rem;
  height: 1.125rem;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  position: static;
}

.c1::before {
  content: '';
  position: absolute;
  color: red;
  top: -2px;
  right: -10px;
}

.c6 {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
  color: #808080;
  height: 1.0625rem;
  margin-top: -5px;
}

<body>
  <div>
    <div
      class="c0"
    >
      <span
        class="c1"
        data-testid="input-description"
      >
         
      </span>
      <div
        class="c2"
      >
        <input
          class="c3 c4"
          data-testid="input"
        />
        <div
          class="c5"
        />
      </div>
      <div
        class="c6"
      >
         
      </div>
    </div>
  </div>
</body>
`;
